<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Master thesis - 8&nbsp; Combining my Seurat and authors Seurat</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<link href="./trajectory_inference.html" rel="next">
<link href="./joining_count_matrices.html" rel="prev">
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./alignment_starsolo.html">Trajectory inference analysis</a></li><li class="breadcrumb-item"><a href="./combining_seurat_objects.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Combining my Seurat and authors Seurat</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Master thesis</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/dariolez/master-thesis-bioinformatics" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./installations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Installations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_download.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data download</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./metadata_inspection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Metadata inspection</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Quality control</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./inspection_authors_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Inspect original paper analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Trajectory inference analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./alignment_starsolo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Aligning scRNA-seq with STARsolo</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./joining_count_matrices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Joining velocity count matrices in a Seurat Object</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./combining_seurat_objects.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Combining my Seurat and authors Seurat</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./trajectory_inference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Trajectory inference</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#comparison-of-seurat-objects-my-object-vs-paper-object" id="toc-comparison-of-seurat-objects-my-object-vs-paper-object" class="nav-link active" data-scroll-target="#comparison-of-seurat-objects-my-object-vs-paper-object"><span class="header-section-number">8.1</span> Comparison of Seurat objects: my object vs paper object</a>
  <ul>
  <li><a href="#comparing-barcodes" id="toc-comparing-barcodes" class="nav-link" data-scroll-target="#comparing-barcodes"><span class="header-section-number">8.1.1</span> Comparing barcodes</a></li>
  <li><a href="#comparing-features" id="toc-comparing-features" class="nav-link" data-scroll-target="#comparing-features"><span class="header-section-number">8.1.2</span> Comparing features</a></li>
  <li><a href="#subsetting-objects-and-merging" id="toc-subsetting-objects-and-merging" class="nav-link" data-scroll-target="#subsetting-objects-and-merging"><span class="header-section-number">8.1.3</span> Subsetting objects and merging</a></li>
  <li><a href="#save-comparison-object" id="toc-save-comparison-object" class="nav-link" data-scroll-target="#save-comparison-object"><span class="header-section-number">8.1.4</span> Save comparison object</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./alignment_starsolo.html">Trajectory inference analysis</a></li><li class="breadcrumb-item"><a href="./combining_seurat_objects.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Combining my Seurat and authors Seurat</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Combining my Seurat and authors Seurat</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="comparison-of-seurat-objects-my-object-vs-paper-object" class="level2" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="comparison-of-seurat-objects-my-object-vs-paper-object"><span class="header-section-number">8.1</span> Comparison of Seurat objects: my object vs paper object</h2>
<p>I compare our analysis with the paper authors one (as requested by Gabriel), with the objective of combining my counts with their object to plot velocity over their UMAP.</p>
<p>Load Seurat saved objects into an clean R session.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>hca_esoph.orig <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"~/TFM/2019.12.31_Meyer/results/oesophagus_ts.rds"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>myesoph.orig <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"~/TFM/2019.12.31_Meyer/results/myesoph_star.rds"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>hca_esoph <span class="ot">&lt;-</span> hca_esoph.orig</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>myesoph <span class="ot">&lt;-</span> myesoph.orig</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="comparing-barcodes" class="level3" data-number="8.1.1">
<h3 data-number="8.1.1" class="anchored" data-anchor-id="comparing-barcodes"><span class="header-section-number">8.1.1</span> Comparing barcodes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Match barcodes between my Seurat object and the paper one</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>bcs_common <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">colnames</span>(hca_esoph), <span class="fu">colnames</span>(myesoph))</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>bcs_no_common <span class="ot">&lt;-</span> <span class="fu">colnames</span>(hca_esoph)[<span class="sc">!</span> <span class="fu">colnames</span>(hca_esoph) <span class="sc">%in%</span> bcs_common]</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the percentage of barcodes retrieved</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co">#data.frame("Number" = c(length(colnames(hca_esoph)), length(rownames(myesoph))), </span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co">#           row.names = c("hca_esoph", "myesoph"))</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(<span class="fu">length</span>(<span class="fu">colnames</span>(hca_esoph)), <span class="fu">length</span>(<span class="fu">colnames</span>(myesoph)))</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>pc_bcs_match <span class="ot">&lt;-</span> (<span class="fu">length</span>(bcs_common) <span class="sc">/</span> <span class="fu">length</span>(<span class="fu">colnames</span>(hca_esoph))) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>pc_bcs_nomatch <span class="ot">&lt;-</span> <span class="fu">length</span>(bcs_no_common) <span class="sc">/</span> <span class="fu">length</span>(<span class="fu">colnames</span>(hca_esoph)) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Matched barcodes: "</span>, <span class="fu">length</span>(bcs_common), <span class="st">" / "</span>, <span class="fu">length</span>(<span class="fu">colnames</span>(hca_esoph)), <span class="st">" ("</span>, pc_bcs_match, <span class="st">"%)</span><span class="sc">\n</span><span class="st">"</span>, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Non-matched barcodes: "</span>, <span class="fu">length</span>(bcs_no_common), <span class="st">" / "</span>, <span class="fu">length</span>(<span class="fu">colnames</span>(hca_esoph)), <span class="st">" ("</span>, pc_bcs_nomatch, <span class="st">"%)</span><span class="sc">\n</span><span class="st">"</span>, <span class="at">sep =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We expect to see a 100% barcode match between the paper’s Seurat object and mine. All of the barcodes in their object should appear in mine, because I basically didn’t filter them. But I get a 77.52% match. The barcodes that appear in the paper object but not in mine come most probably from the 3 samples that I didn’t include in the analysis. I’ll check this.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(<span class="fu">gsub</span>(<span class="st">"[ACGT]{16}"</span>, <span class="st">""</span>, <span class="fu">colnames</span>(hca_esoph)[<span class="sc">!</span> <span class="fu">colnames</span>(hca_esoph) <span class="sc">%in%</span> bcs_common]))</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>pc_bcs <span class="ot">&lt;-</span> (<span class="fu">length</span>(bcs_common) <span class="sc">/</span> <span class="fu">length</span>(<span class="fu">grep</span>(<span class="st">".*74136(20|21|22)"</span>, <span class="fu">colnames</span>(hca_esoph), <span class="at">invert =</span> <span class="cn">TRUE</span>))) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Matched barcodes:"</span>, pc_bcs, <span class="st">"%</span><span class="sc">\n</span><span class="st">"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>All of the barcodes that don’t match come from the 3 samples I removed from the analysis for their bad base quality. If I remove these barcodes/cells when calculating the match percentage between my object and the paper’s, I get 100% match (as expected). It’s surprising to see that the authors didn’t remove this samples from the analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check barcodes names are identical</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">sort</span>(<span class="fu">grep</span>(<span class="st">".*74136(20|21|22)"</span>, <span class="fu">colnames</span>(hca_esoph), <span class="at">invert =</span> <span class="cn">TRUE</span>, <span class="at">value =</span> <span class="cn">TRUE</span>)) <span class="sc">==</span> <span class="fu">sort</span>(<span class="fu">colnames</span>(myesoph)[<span class="fu">colnames</span>(myesoph) <span class="sc">%in%</span> bcs_common]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="comparing-features" class="level3" data-number="8.1.2">
<h3 data-number="8.1.2" class="anchored" data-anchor-id="comparing-features"><span class="header-section-number">8.1.2</span> Comparing features</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check common features:</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>genes_common <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">rownames</span>(hca_esoph), <span class="fu">rownames</span>(myesoph))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>genes_no_common <span class="ot">&lt;-</span> <span class="fu">rownames</span>(hca_esoph)[<span class="sc">!</span> <span class="fu">rownames</span>(hca_esoph) <span class="sc">%in%</span> genes_common]</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(<span class="fu">length</span>(<span class="fu">rownames</span>(hca_esoph)), <span class="fu">length</span>(<span class="fu">rownames</span>(myesoph)))</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>pc_genes_match <span class="ot">&lt;-</span> <span class="fu">length</span>(genes_common) <span class="sc">/</span> <span class="fu">length</span>(<span class="fu">rownames</span>(hca_esoph)) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>pc_genes_nomatch <span class="ot">&lt;-</span> <span class="fu">length</span>(genes_no_common) <span class="sc">/</span> <span class="fu">length</span>(<span class="fu">rownames</span>(hca_esoph)) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Matched genes: "</span>, <span class="fu">length</span>(genes_common), <span class="st">" / "</span>, <span class="fu">length</span>(<span class="fu">rownames</span>(hca_esoph)), <span class="st">" ("</span>, pc_genes_match, <span class="st">"%)</span><span class="sc">\n</span><span class="st">"</span>, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Non-matched genes: "</span>, <span class="fu">length</span>(genes_no_common), <span class="st">" / "</span>, <span class="fu">length</span>(<span class="fu">rownames</span>(hca_esoph)), <span class="st">" ("</span>, pc_genes_nomatch, <span class="st">"%)</span><span class="sc">\n</span><span class="st">"</span>, <span class="at">sep =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are 27.85% of gene names unmatched between objects. This percentage is quite high, I’m going to use the Ensembl gene IDs stored in the <code>meta.features</code> slot of the assays to see if I can match more.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate a new metadata column without the version numbers from the Ensembl IDs in myesoph assays</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>myesoph<span class="sc">@</span>assays<span class="sc">$</span>spliced<span class="sc">@</span>meta.features[[<span class="st">"ensembl_id_noversion"</span>]] <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(myesoph<span class="sc">@</span>assays<span class="sc">$</span>spliced<span class="sc">@</span>meta.features<span class="sc">$</span>ensembl_id, <span class="st">"</span><span class="sc">\\</span><span class="st">.</span><span class="sc">\\</span><span class="st">d+"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>myesoph<span class="sc">@</span>assays<span class="sc">$</span>unspliced<span class="sc">@</span>meta.features[[<span class="st">"ensembl_id_noversion"</span>]] <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(myesoph<span class="sc">@</span>assays<span class="sc">$</span>unspliced<span class="sc">@</span>meta.features<span class="sc">$</span>ensembl_id, <span class="st">"</span><span class="sc">\\</span><span class="st">.</span><span class="sc">\\</span><span class="st">d+"</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Match my merged object and the paper one</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co"># I take the first column of the paper object, because all are the same, and the spliced column</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="co"># from my merged object because the spliced and unspliced ensembl IDs are identical</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>ens_genes_common <span class="ot">&lt;-</span> <span class="fu">intersect</span>(hca_esoph<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.features[[<span class="dv">1</span>]], myesoph<span class="sc">@</span>assays<span class="sc">$</span>spliced<span class="sc">@</span>meta.features<span class="sc">$</span>ensembl_id_noversion)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>ens_genes_no_common <span class="ot">&lt;-</span> hca_esoph<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.features[[<span class="dv">1</span>]][<span class="sc">!</span> hca_esoph<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.features[[<span class="dv">1</span>]] <span class="sc">%in%</span> ens_genes_common]</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(<span class="fu">length</span>(<span class="fu">rownames</span>(hca_esoph)), <span class="fu">length</span>(<span class="fu">rownames</span>(myesoph)))</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>pc_ens_genes_match <span class="ot">&lt;-</span> <span class="fu">length</span>(ens_genes_common) <span class="sc">/</span> <span class="fu">length</span>(<span class="fu">rownames</span>(hca_esoph)) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>pc_ens_genes_nomatch <span class="ot">&lt;-</span> <span class="fu">length</span>(ens_genes_no_common) <span class="sc">/</span> <span class="fu">length</span>(<span class="fu">rownames</span>(hca_esoph)) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Matched genes: "</span>, <span class="fu">length</span>(ens_genes_common), <span class="st">" / "</span>, <span class="fu">length</span>(<span class="fu">rownames</span>(hca_esoph)), <span class="st">" ("</span>, pc_ens_genes_match, <span class="st">"%)</span><span class="sc">\n</span><span class="st">"</span>, <span class="at">sep =</span> <span class="st">""</span>)</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Non-matched genes: "</span>, <span class="fu">length</span>(ens_genes_no_common), <span class="st">" / "</span>, <span class="fu">length</span>(<span class="fu">rownames</span>(hca_esoph)), <span class="st">" ("</span>, pc_ens_genes_nomatch, <span class="st">"%)</span><span class="sc">\n</span><span class="st">"</span>, <span class="at">sep =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Using Ensembl IDs 98.31% of genes match between objects. There are 410 Ensembl IDs (1.69%) in the paper’s object that don’t appear in my object. I check if some of these genes actually match by symbol but not ID, because of differences between the reference genomes used.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get non matched gene symbols</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>ens_genes_no_common_symbol <span class="ot">&lt;-</span> <span class="fu">rownames</span>(hca_esoph)[<span class="sc">!</span> hca_esoph<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.features[[<span class="dv">1</span>]] <span class="sc">%in%</span> ens_genes_common]</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># See if all the non-match Ensembl IDs are included in the non-matched symbols</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We would expect that all the genes whose Ensembl IDs don't match, also don't match by symbol</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>no_common_ens_and_symbol <span class="ot">&lt;-</span> <span class="fu">intersect</span>(genes_no_common, ens_genes_no_common_symbol)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">length</span>(ens_genes_no_common))</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">length</span>(no_common_ens_and_symbol))</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Get genes whose Ensembl ID doesn't match but the symbol does</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>ens_genes_no_common[ens_genes_no_common_symbol <span class="sc">%in%</span> genes_common]</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>ens_genes_no_common_symbol[ens_genes_no_common_symbol <span class="sc">%in%</span> genes_common]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are 33 genes that match by symbol but not by Ensembl ID. This is probably due to changes and deprecations in Ensembl IDs between the reference genome the paper authors used and the one that I used. Keeping these genes would improve the match between objects in 0.14%. Such a little improvement isn’t worth the effort that would take to add them to the match. I looked if there are any noticeably important genes, and it doesn’t seem so. I’ll skip these genes.</p>
</section>
<section id="subsetting-objects-and-merging" class="level3" data-number="8.1.3">
<h3 data-number="8.1.3" class="anchored" data-anchor-id="subsetting-objects-and-merging"><span class="header-section-number">8.1.3</span> Subsetting objects and merging</h3>
<p>Subset the paper’s object and my merged object so that they have the same barcodes (columns) and genes (rows).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># RESTRICT TO COMMON BARCODES</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove the samples that I didn't include from the paper's object</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>hca_esoph <span class="ot">&lt;-</span> <span class="fu">subset</span>(hca_esoph, <span class="at">cells =</span> <span class="fu">grep</span>(<span class="st">".*74136(20|21|22)"</span>, <span class="fu">colnames</span>(hca_esoph), <span class="at">invert =</span> <span class="cn">TRUE</span>, <span class="at">value =</span> <span class="cn">TRUE</span>))</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove the barcodes that don't appear in the paper's object</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>myesoph <span class="ot">&lt;-</span> <span class="fu">subset</span>(myesoph, <span class="at">cells =</span> bcs_common)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># RESTRICT TO COMMON FEATURES</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>hca_esoph <span class="ot">&lt;-</span> <span class="fu">subset</span>(hca_esoph, <span class="at">features =</span> <span class="fu">rownames</span>(hca_esoph)[hca_esoph<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.features[[<span class="st">"gene.ids-HCATisStab7413619"</span>]] <span class="sc">%in%</span> ens_genes_common])</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>myesoph <span class="ot">&lt;-</span> <span class="fu">subset</span>(myesoph, <span class="at">features =</span> <span class="fu">rownames</span>(myesoph)[myesoph<span class="sc">@</span>assays<span class="sc">$</span>spliced<span class="sc">@</span>meta.features<span class="sc">$</span>ensembl_id_noversion <span class="sc">%in%</span> ens_genes_common])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check that the Seurat objects have columns and rows ordered equally, before joining the objects.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check that the barcodes are identical and ordered equally</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">colnames</span>(hca_esoph) <span class="sc">==</span> <span class="fu">colnames</span>(myesoph))</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Check that the features/genes are identical and ordered equally</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">rownames</span>(hca_esoph) <span class="sc">==</span> <span class="fu">rownames</span>(myesoph))  <span class="co"># this will be FALSE (different HGNC symbols)</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(hca_esoph[[<span class="st">"RNA"</span>]]<span class="sc">@</span>meta.features[[<span class="dv">1</span>]] <span class="sc">==</span> myesoph[[<span class="st">"spliced"</span>]]<span class="sc">@</span>meta.features<span class="sc">$</span>ensembl_id_noversion)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Barcodes (columns) are ordered identically. But features (rows) don’t match between objects, as expected. Each one has HGNC symbols, which are not guaranteed to match, and genes are most probably ordered differently. To add my count matrices to the paper’s object I need to match Ensembl IDs and re-order the rows (genes) in my merged object.</p>
<p>Reorder genes and add my object’s assays to the paper’s object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Match rows in my merged object with the papers object to reorder them before merging</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>hca_myspliced_match <span class="ot">&lt;-</span> <span class="fu">match</span>(hca_esoph<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.features[[<span class="dv">1</span>]], myesoph<span class="sc">@</span>assays<span class="sc">$</span>spliced<span class="sc">@</span>meta.features<span class="sc">$</span>ensembl_id_noversion)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>hca_myunspliced_match <span class="ot">&lt;-</span> <span class="fu">match</span>(hca_esoph<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.features[[<span class="dv">1</span>]], myesoph<span class="sc">@</span>assays<span class="sc">$</span>unspliced<span class="sc">@</span>meta.features<span class="sc">$</span>ensembl_id_noversion)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Check that we will store the genes correctly</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(myesoph[[<span class="st">"spliced"</span>]]<span class="sc">@</span>meta.features<span class="sc">$</span>ensembl_id_noversion[hca_myspliced_match] <span class="sc">==</span> hca_esoph<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.features[[<span class="dv">1</span>]])</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(myesoph[[<span class="st">"unspliced"</span>]]<span class="sc">@</span>meta.features<span class="sc">$</span>ensembl_id_noversion[hca_myunspliced_match] <span class="sc">==</span> hca_esoph<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>meta.features[[<span class="dv">1</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add my assays to the papers object</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>hca_esoph[[<span class="st">"spliced"</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(myesoph<span class="sc">@</span>assays<span class="sc">$</span>spliced<span class="sc">@</span>counts[hca_myspliced_match, ])</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>hca_esoph[[<span class="st">"unspliced"</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(myesoph<span class="sc">@</span>assays<span class="sc">$</span>unspliced<span class="sc">@</span>counts[hca_myunspliced_match, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you don’t want to use the Ensembl IDs, and match only HGCN symbols, do the following.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>hca_esoph[[<span class="st">"spliced"</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(myesoph<span class="sc">@</span>assays<span class="sc">$</span>spliced<span class="sc">@</span>counts)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>hca_esoph[[<span class="st">"unspliced"</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(myesoph<span class="sc">@</span>assays<span class="sc">$</span>unspliced<span class="sc">@</span>counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Compare counts from the papers assay with my assays.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>To calculate <code>nCounts</code> manually do: <code>Matrix::colSums(seuratobject@assays$name@counts)</code> To calculate <code>nFeature</code> manually do: ``</p>
<p>If you try to do <code>Matrix::colSums(seuratobject[["name"]])</code>, it will sum the values in the slot <code>data</code> that is were normalization values are stored, and the results will be all wrong.</p>
</div>
</div>
<p>Compare counts for spliced and unspliced separately.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare total reads per barcode (nCount) for RNA vs. Spliced &amp; Unspliced</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"results/nCount_comparison.pdf"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>axis_max <span class="ot">&lt;-</span> <span class="fu">round_any</span>(<span class="fu">max</span>(hca_esoph<span class="sc">$</span>nCount_RNA, hca_esoph<span class="sc">$</span>nCount_spliced), <span class="dv">10</span>, ceiling)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hca_esoph<span class="sc">$</span>nCount_RNA, hca_esoph<span class="sc">$</span>nCount_spliced, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max))</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hca_esoph<span class="sc">$</span>nCount_RNA, hca_esoph<span class="sc">$</span>nCount_unspliced, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max))</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(hca_esoph<span class="sc">$</span>nCount_RNA)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(hca_esoph<span class="sc">$</span>nCount_spliced)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Compare features for spliced and unspliced separately.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare number of genes per barcode (nFeature) for RNA vs Spliced &amp; Unspliced</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"results/nFeature_comparison.pdf"</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>axis_max <span class="ot">&lt;-</span> <span class="fu">round_any</span>(<span class="fu">max</span>(hca_esoph<span class="sc">$</span>nFeature_RNA, hca_esoph<span class="sc">$</span>nFeature_spliced), <span class="dv">10</span>, ceiling)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hca_esoph<span class="sc">$</span>nFeature_RNA, hca_esoph<span class="sc">$</span>nFeature_spliced, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max))</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hca_esoph<span class="sc">$</span>nFeature_RNA, hca_esoph<span class="sc">$</span>nFeature_unspliced, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max))</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(hca_esoph<span class="sc">$</span>nFeature_RNA)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(hca_esoph<span class="sc">$</span>nFeature_spliced)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Compare counts and features for spliced and unspliced of my object combined against the paper object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare nCount and nFeature for RNA vs Spliced+Unspliced</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="at">file=</span><span class="st">"results/nCount_nFeature_comparison.pdf"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>axis_max <span class="ot">&lt;-</span> <span class="fu">round_any</span>(<span class="fu">max</span>(hca_esoph<span class="sc">$</span>nCount_RNA, hca_esoph<span class="sc">$</span>nCount_spliced), <span class="dv">10</span>, ceiling)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hca_esoph<span class="sc">$</span>nCount_RNA, (hca_esoph<span class="sc">$</span>nCount_spliced <span class="sc">+</span> hca_esoph<span class="sc">$</span>nCount_unspliced), <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max))</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="st">"nCount"</span>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>axis_max <span class="ot">&lt;-</span> <span class="fu">round_any</span>(<span class="fu">max</span>(hca_esoph<span class="sc">$</span>nFeature_RNA, hca_esoph<span class="sc">$</span>nFeature_spliced), <span class="dv">10</span>, ceiling)</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hca_esoph<span class="sc">$</span>nFeature_RNA, (hca_esoph<span class="sc">$</span>nFeature_spliced <span class="sc">+</span> hca_esoph<span class="sc">$</span>nFeature_unspliced), <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, axis_max))</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="fu">c</span>(<span class="dv">0</span>, axis_max), <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="st">"nFeature"</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(hca_esoph<span class="sc">$</span>nFeature_RNA)</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(hca_esoph<span class="sc">$</span>nFeature_spliced)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Calculate difference between my reads and the paper reads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>((hca_esoph<span class="sc">$</span>nCount_spliced <span class="sc">+</span> hca_esoph<span class="sc">$</span>nCount_unspliced) <span class="sc">-</span> hca_esoph<span class="sc">$</span>nCount_RNA)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>((hca_esoph<span class="sc">$</span>nFeature_spliced <span class="sc">+</span> hca_esoph<span class="sc">$</span>nFeature_unspliced) <span class="sc">-</span> hca_esoph<span class="sc">$</span>nFeature_RNA)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the difference between my object and the paper's in percentage value</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># A result of 0%, would be that both objects are identical, positive values mean my object has more counts</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>(( <span class="fu">sum</span>(hca_esoph<span class="sc">$</span>nCount_spliced <span class="sc">+</span> hca_esoph<span class="sc">$</span>nCount_unspliced) <span class="sc">/</span> <span class="fu">sum</span>(hca_esoph<span class="sc">$</span>nCount_RNA) ) <span class="sc">-</span> <span class="dv">1</span> ) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>(( <span class="fu">sum</span>(hca_esoph<span class="sc">$</span>nFeature_spliced <span class="sc">+</span> hca_esoph<span class="sc">$</span>nFeature_unspliced) <span class="sc">/</span> <span class="fu">sum</span>(hca_esoph<span class="sc">$</span>nFeature_RNA) ) <span class="sc">-</span> <span class="dv">1</span> ) <span class="sc">*</span> <span class="dv">100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Judging by the plots comparing the sum of spliced+unspliced against the paper counts and the summary statistics, we conclude that:</p>
<ul>
<li>The paper object has more reads per barcode than my object in average (nCount). My object has -11.5% reads less than the paper object. This could be because of the UMI deduplication strategy used, I used the default “1MM All” of STARsolo, but if you use “Exact”, you could recover more reads per barcode. But there are other possible explanations.</li>
<li>My object has more genes per barcode than the paper object in average (nFeature). My object has +9.33% genes per barcode more than the paper object. This could be happening because I haven’t done any filtering at all compared to paper object. But again, there could be other explanations.</li>
</ul>
</section>
<section id="save-comparison-object" class="level3" data-number="8.1.4">
<h3 data-number="8.1.4" class="anchored" data-anchor-id="save-comparison-object"><span class="header-section-number">8.1.4</span> Save comparison object</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(hca_esoph, <span class="at">file=</span><span class="st">"results/hca_myesoph.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./joining_count_matrices.html" class="pagination-link  aria-label=" &lt;span="" velocity="" count="" matrices="" in="" a="" seurat="" object&lt;="" span&gt;"="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Joining velocity count matrices in a Seurat Object</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./trajectory_inference.html" class="pagination-link" aria-label="<span class='chapter-number'>9</span>&nbsp; <span class='chapter-title'>Trajectory inference</span>">
        <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Trajectory inference</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>